File: backend/internal/service/menu.go
package service

import (
	"keystonego/internal/model"
	"keystonego/internal/repository"
	"keystonego/pkg/utils"

	"gorm.io/gorm"
)

// MenuService 是菜单管理业务逻辑服务。
type MenuService struct {
	menuRepo *repository.MenuRepository
	roleRepo *repository.RoleRepository
	db       *gorm.DB
}

// NewMenuService 创建菜单服务（Wire Provider）。
func NewMenuService(menuRepo *repository.MenuRepository, roleRepo *repository.RoleRepository, db *gorm.DB) *MenuService {
	return &MenuService{menuRepo: menuRepo, roleRepo: roleRepo, db: db}
}

// ListTree 返回菜单树形结构（调用 utils.BuildMenuTree 将扁平列表转为嵌套树）。
func (s *MenuService) ListTree() ([]*utils.MenuTreeResp, error) {
	menus, err := s.menuRepo.ListTree()
	if err != nil {
		return nil, err
	}
	return utils.BuildMenuTree(menus), nil
}

// GetByID 按 ID 查询单个菜单。
func (s *MenuService) GetByID(id uint64) (*model.Menu, error) {
	return s.menuRepo.FindByID(id)
}

// Create 创建菜单。
func (s *MenuService) Create(menu *model.Menu) error {
	return s.menuRepo.Create(menu)
}

// Update 更新菜单（部分更新，仅更新有值的字段，Hidden 直接覆盖）。
func (s *MenuService) Update(id uint64, menu *model.Menu) error {
	existing, err := s.menuRepo.FindByID(id)
	if err != nil {
		return err
	}
	if menu.Name != "" {
		existing.Name = menu.Name
	}
	if menu.Path != "" {
		existing.Path = menu.Path
	}
	if menu.Component != "" {
		existing.Component = menu.Component
	}
	if menu.Icon != "" {
		existing.Icon = menu.Icon
	}
	if menu.Sort != 0 {
		existing.Sort = menu.Sort
	}
	if menu.Type != 0 {
		existing.Type = menu.Type
	}
	existing.Hidden = menu.Hidden
	existing.Perms = menu.Perms
	return s.menuRepo.Update(existing)
}

// Delete 删除菜单。
func (s *MenuService) Delete(id uint64) error {
	return s.menuRepo.Delete(id)
}

// UpdateStatus 更新菜单启用/禁用状态。
func (s *MenuService) UpdateStatus(id uint64, status int8) error {
	return s.menuRepo.UpdateStatus(id, status)
}

// GetUserMenus 获取指定用户可访问的菜单树（通过角色-菜单关联）。
// 查询结果为空时兜底返回全部菜单，避免新用户看到空侧边栏。
func (s *MenuService) GetUserMenus(userID uint64) ([]*utils.MenuTreeResp, error) {
	menus, err := s.menuRepo.GetUserMenus(userID)
	if err != nil {
		return nil, err
	}
	if len(menus) == 0 {
		// 兜底：用户无角色分配时返回所有菜单
		menus, _ = s.menuRepo.ListTree()
	}
	return utils.BuildMenuTree(menus), nil
}

// InitDefaultMenus 初始化系统默认菜单结构（使用固定 ID + FirstOrCreate 确保幂等）。
// 菜单结构：
//
//	首页 (/dashboard)
//	系统管理 (/system)
//	  ├── 用户管理 (/system/users)
//	  ├── 角色管理 (/system/roles)
//	  ├── 菜单管理 (/system/menus)
//	  ├── 权限管理 (/system/permissions)
//	  └── 系统监控 (/system/monitor)
//
// 初始化完成后自动将所有菜单分配给 admin 角色（ID=1）。
func (s *MenuService) InitDefaultMenus() error {
	menus := []model.Menu{
		{BaseModel: model.BaseModel{ID: 1}, Name: "首页", Path: "/dashboard", Component: "dashboard/index", Icon: "HomeFilled", Sort: 1, Type: 2, Status: 1},
		{BaseModel: model.BaseModel{ID: 2}, Name: "系统管理", Path: "/system", Icon: "Setting", Sort: 2, Type: 1, Status: 1},
		{BaseModel: model.BaseModel{ID: 3}, ParentID: 2, Name: "用户管理", Path: "/system/users", Component: "system/user/index", Icon: "User", Sort: 1, Type: 2, Status: 1},
		{BaseModel: model.BaseModel{ID: 4}, ParentID: 2, Name: "角色管理", Path: "/system/roles", Component: "system/role/index", Icon: "UserFilled", Sort: 2, Type: 2, Status: 1},
		{BaseModel: model.BaseModel{ID: 5}, ParentID: 2, Name: "菜单管理", Path: "/system/menus", Component: "system/menu/index", Icon: "Menu", Sort: 3, Type: 2, Status: 1},
		{BaseModel: model.BaseModel{ID: 6}, ParentID: 2, Name: "权限管理", Path: "/system/permissions", Component: "system/permission/index", Icon: "Key", Sort: 4, Type: 2, Status: 1},
		{BaseModel: model.BaseModel{ID: 7}, ParentID: 2, Name: "系统监控", Path: "/system/monitor", Component: "system/monitor/index", Icon: "Monitor", Sort: 5, Type: 2, Status: 1},
	}

	// FirstOrCreate 确保重复启动不会创建重复菜单
	for _, m := range menus {
		s.db.Where("id = ?", m.ID).FirstOrCreate(&m)
	}

	// 分配所有菜单给 admin 角色
	var allMenuIDs []uint64
	for i := 1; i <= 7; i++ {
		allMenuIDs = append(allMenuIDs, uint64(i))
	}
	s.roleRepo.AssignRoleMenus(1, allMenuIDs)

	return nil
}

File: frontend/src/api/menu.ts
/**
 * 菜单 API 模块 — 菜单树查询、CRUD、用户菜单。
 *
 * MenuItem 接口对应后端 MenuTreeResp 结构：
 *   - type: 'M'=目录, 'C'=菜单, 'F'=按钮
 *   - hidden: 是否在侧边栏隐藏（与后端 Visible 字段逻辑取反）
 *   - children: 子菜单（树形结构）
 */
import request from '@/utils/request'

/** 菜单项接口定义 */
export interface MenuItem {
  id: number
  parent_id: number
  name: string
  path: string
  component?: string   // 前端组件路径（仅 type=C 时有效）
  icon?: string        // Element Plus 图标名
  sort: number
  type: string         // M=目录, C=菜单, F=按钮
  perms?: string       // 权限标识（仅 type=F 时有效）
  status: number       // 1=启用, 0=禁用
  hidden: boolean      // 是否在侧边栏隐藏
  children?: MenuItem[]
  created_at?: string
  updated_at?: string
}

/** 获取菜单列表（扁平结构） */
export function getMenus() {
  return request.get('/menus')
}

/** 获取菜单树（嵌套结构） */
export function getMenuTree() {
  return request.get('/menus/tree')
}

/** 按 ID 查询单个菜单 */
export function getMenu(id: number) {
  return request.get(`/menus/${id}`)
}

/** 创建菜单 */
export function createMenu(data: Partial<MenuItem>) {
  return request.post('/menus', data)
}

/** 更新菜单（部分更新） */
export function updateMenu(id: number, data: Partial<MenuItem>) {
  return request.put(`/menus/${id}`, data)
}

/** 删除菜单 */
export function deleteMenu(id: number) {
  return request.delete(`/menus/${id}`)
}

/** 更新菜单启用/禁用状态 */
export function updateMenuStatus(id: number, status: number) {
  return request.put(`/menus/${id}/status`, { status })
}

/** 获取当前用户可访问的菜单树（按角色过滤） */
export function getUserMenus() {
  return request.get('/user/menus')
}
